primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+    pub fn expect_ok(self) -> Result<(F, P), PrimitivesError> {
+        match self {
+            LookupResult::Ok(x, proof) => Ok((x, proof)),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected Ok, found NotInMemory".to_string(),
+            )),
+            LookupResult::EmptyLeaf(_) => Err(PrimitivesError::InternalError(
+                "Expected Ok, found EmptyLeaf".to_string(),
+            )),
+        }
+    }
+
+    /// Assert the lookup result is NotInMemory. Return a non-membership proof.
+    pub fn expect_empty(self) -> Result<N, PrimitivesError> {
+        match self {
+            LookupResult::EmptyLeaf(n) => Ok(n),
+            LookupResult::Ok(..) => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found element".to_string(),
+            )),
+            LookupResult::NotInMemory => Err(PrimitivesError::InternalError(
+                "Expected EmptyLeaf, found NotInMemory".to_string(),
+            )),
+        }
+    }
+}
+
+/// An element of a Merkle tree.
+pub trait Element: Clone + Eq + PartialEq {}
+impl<T: Clone + Eq + PartialEq> Element for T {}
+
+/// An index type of a leaf in a Merkle tree.
+pub trait Index: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath {}
+impl<T: Debug + Eq + PartialEq + Ord + PartialOrd + Clone + ToTraversalPath> Index for T {}
+
+/// An internal node value type in a Merkle tree.
+pub trait NodeValue: Default + Eq + PartialEq + Copy + Clone + Debug {}
+impl<T: Default + Eq + PartialEq + Copy + Clone + Debug> NodeValue for T {}
+
+/// Merkle tree hash function
+pub trait DigestAlgorithm<E, I, T>
+where
+    E: Element,
+    I: Index,
+    T: NodeValue,
+{
+    // Possible improvement: adding digest_element() and digest_index()
+    /// Digest a list of values
+    fn digest(data: &[T]) -> T;
+
+    /// Digest a leaf (an indexed element)
+    fn digest_leaf(pos: &I, elem: &E) -> T;
+}
+
+/// An trait for Merkle tree index type.
+pub trait ToTraversalPath {
+    /// Convert the given index to a vector of branch indices given tree height
+    /// and arity.
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize>;
+}
+
+impl<F: Into<BigUint> + Clone> ToTraversalPath for F {
+    fn to_traverse_path(&self, height: usize, arity: usize) -> Vec<usize> {
+        let mut pos: BigUint = <F as Into<BigUint>>::into(self.clone());
+        let mut ret = vec![];
+        for _i in 0..height {
+            ret.push((&pos % (arity as u64)).to_usize().unwrap());
+            pos /= arity as u64;
+        }
+        ret
+    }
+}
+
+/// A merkle commitment consists a root hash value, a tree height and number of
+/// leaves
+#[derive(Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+pub struct MerkleCommitment<T: NodeValue> {
+    /// Root of a tree
+    pub root_value: T,
+    /// Height of a tree
+    pub height: usize,
+    /// Number of leaves in the tree
+    pub num_leaves: u64,
+}
```

Comment:
This seems like it would be very specific to an implementation. If the intent is for the trait below to be implementable for a variety of accumulators, I think this should be replaced by a `trait MerkleCommitment<T: NodeValue>` and an associated type on `MerkleTreeScheme` like `type Commitment: MerkleCommitment`. Interface-wise, I don't think the exact structure of this type is important for users, as long as it satisfies the following properties:
* concise (e.g. `Copy`)
* can be used to identify a tree (`Eq`, `Ord`, `Hash`)
* exposes the size of the tree (`num_leaves`, but we should expose it via a trait function with a more abstract name like `size` or something)

Reply:
Now I have a trait named `MerkleCommitment` and a struct implements it called `MerkleTreeCommitment`.
It still may bring confusions to users. Are there any naming suggestion?